File: Documentation/5.2/Samples/nodejs/documentExtensions/counters/indexingCounters.js
import {
    AbstractCountersIndexCreationTask, AbstractRawJavaScriptCountersIndexCreationTask,
    DocumentStore
} from "ravendb";

const store = new DocumentStore();
const session = store.openSession();

//region index_1
export class MyCounterIndex  extends AbstractCountersIndexCreationTask {
    constructor() {
        super();
        this.map = `from counter in docs.counters select new {
                Likes = counter.Value,
                Name = counter.Name,
                User = counter.DocumentId
                }`;
    }
}
//endregion

let map,reduce;
//region javaScriptIndexCreationTask
class CsharpCountersIndexCreationTask  extends AbstractCountersIndexCreationTask {
    public constructor() {
        super();

        this.map = map;

        this.reduce = reduce;
    }
}
//endregion

//region index_3
class MyCounterIndex extends AbstractRawJavaScriptCountersIndexCreationTask {
    public constructor() {
        super();

        this.maps.add(
            "counters.map('Companies', 'HeartRate', function (counter) {\n" +
            "return {\n" +
            "    heartBeat: counter.Value,\n" +
            "    name: counter.Name,\n" +
            "    user: counter.DocumentId\n" +
            "};\n" +
            "})"
        );
    }
}
//endregion

//region index_0
class Companies_ByCounterNames extends AbstractCountersIndexCreationTask{
    constructor() {
        super();
        this.map = "docs.Companies.Select(e => new {\n"+
            "e = e,\n"+
        "    counterNames = this.CounterNamesFor(e)\n"+
        "}).Select(this0 => new {\n"+
        "    CounterNames = Enumerable.ToArray(this0.counterNames)\n"+
        "})"
    }
}
//endregion


//region query1
const companies = session
    .query({index: "Companies_ByCounterNames"})
    .containsAny("counterNames", ["Likes"])
    .all();
//endregion

File: Documentation/5.1/Samples/java/src/test/java/net/ravendb/Indexes/IndexingCounters.java
package net.ravendb.Indexes;

import com.google.common.collect.Lists;
import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.session.IDocumentSession;
import net.ravendb.northwind.Company;

import java.util.List;

public class IndexingCounters {
    private interface IFoo {
        //region syntax
        List<String> CounterNamesFor(Object doc);
        //endregion
    }

    //region index
    public static class Companies_ByCounterNames extends AbstractIndexCreationTask {
        public Companies_ByCounterNames() {
            map = "from e in docs.Employees\n" +
                "let counterNames = CounterNamesFor(e)\n" +
                "select new\n" +
                "{\n" +
                "   counterNames = counterNames.ToArray()\n" +
                "}";
        }
    }
    //endregion

    public void sample() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region query1
                // return all companies that have 'Likes' counter
                List<Company> companies = session
                    .query(Company.class, Companies_ByCounterNames.class)
                    .containsAny("counterNames", Lists.newArrayList("Likes"))
                    .toList();
                //endregion
            }
        }
    }

    //region
    protected void addMap(String map)
    //endregion
    {
        if (map == null) {
            throw new IllegalArgumentException("Map cannot be null");
        }
        maps.add(map);
    }

    //region index_1
    public static class MyCounterIndex extends AbstractCountersIndexCreationTask {
        public MyCounterIndex() {
            map = "counters.Companies.HeartRate.Select(counter => new {\n" +
                    "    heartBeat = counter.Value,\n" +
                    "    name = counter.Name,\n" +
                    "    user = counter.DocumentId\n" +
                    "})";
        }
    }
    //endregion


}
